orttraining/orttraining/python/training/ortmodule/torch_cpp_extensions/aten_op_executor/aten_op_executor.cc

```diff
@@ -6,11 +6,107 @@
 #include <unordered_map>
 #include <vector>
 
+template <typename T>
+c10::IValue ToIValue(const DLManagedTensor* dlpack, bool is_optional) {
+  TORCH_INTERNAL_ASSERT((dlpack->dl_tensor.ndim == 0 && dlpack->dl_tensor.shape == nullptr) ||
+                        (dlpack->dl_tensor.ndim == 1 && dlpack->dl_tensor.shape[0] == 1));
+  T value = *reinterpret_cast<const T*>(dlpack->dl_tensor.data);
+  return is_optional ? c10::IValue(c10::optional<T>(value)) : c10::IValue(value);
+}
+
+template <typename T>
+c10::IValue ToListIValue(const DLManagedTensor* dlpack, bool is_optional) {
+  TORCH_INTERNAL_ASSERT(dlpack->dl_tensor.ndim == 1);
+  const T* p_data = reinterpret_cast<const T*>(dlpack->dl_tensor.data);
+  c10::List<T> list_value;
+  size_t len = static_cast<size_t>(dlpack->dl_tensor.shape[0]);
+  for (size_t i = 0; i < len; i++) {
+    list_value.emplace_back(p_data[i]);
+  }
+  return is_optional ? c10::IValue(c10::optional<c10::List<T>>(list_value)) : c10::IValue(list_value);
+}
+
+c10::IValue Int64ToBoolIValue(const DLManagedTensor* dlpack, bool is_list, bool is_optional) {
+  if (is_list) {
+    TORCH_INTERNAL_ASSERT(dlpack->dl_tensor.ndim == 1);
+    const int64_t* p_data = reinterpret_cast<const int64_t*>(dlpack->dl_tensor.data);
+    c10::List<bool> list_value;
+    size_t len = static_cast<size_t>(dlpack->dl_tensor.shape[0]);
+    for (size_t i = 0; i < len; i++) {
+      list_value.emplace_back(static_cast<bool>(p_data[i]));
+    }
+    return is_optional ? c10::IValue(c10::optional<c10::List<bool>>(list_value)) : c10::IValue(list_value);
+  }
+
+  TORCH_INTERNAL_ASSERT((dlpack->dl_tensor.ndim == 0 && dlpack->dl_tensor.shape == nullptr) ||
+                        (dlpack->dl_tensor.ndim == 1 && dlpack->dl_tensor.shape[0] == 1));
+  bool value = static_cast<bool>(*reinterpret_cast<const int64_t*>(dlpack->dl_tensor.data));
+  return is_optional ? c10::IValue(c10::optional<bool>(value)) : c10::IValue(value);
+}
+
 struct ATenOperator {
   std::shared_ptr<torch::jit::Operator> op;
   size_t argument_size;
+  std::vector<c10::TypeKind> elem_kinds;
+  std::vector<bool> is_list_arguments;
   std::vector<bool> is_optional_arguments;
+  std::vector<c10::optional<c10::IValue>> default_values;
   size_t return_size;
+
+  c10::IValue ToIValueArgument(const DLManagedTensor* dlpack, size_t index) const {
+    TORCH_INTERNAL_ASSERT(index < argument_size);
+    bool is_optional = is_optional_arguments[index];
+    TORCH_INTERNAL_ASSERT(dlpack || is_optional || default_values[index]);
+    if (!dlpack) {
+      if (is_optional) {
+        // Optional argument always has no default value.
+        return c10::IValue(c10::nullopt);
+      }
+
+      return *default_values[index];
```

Comment:
I didn't check c10:optional class signature, but if the "*" is expected?

Reply:
PyTorch code also uses in such way.